server/common.ts

```diff
@@ -406,14 +418,49 @@ export function authenticateWithReject(request: express.Request, response: expre
 	}
 }
 // For directly user facing endpoints
-export function authenticateWithRedirect(request: express.Request, response: express.Response, next: express.NextFunction) {
+export async function authenticateWithRedirect(request: express.Request, response: express.Response, next: express.NextFunction) {
 	if (!request.isAuthenticated()) {
-		response.redirect("/login");
+		await redirectToLogin(request, response);
 	}
 	else {
 		next();
 	}
 }
+
+export function getExternalPort(request: express.Request): number {
+	function defaultPort(): number {
+		// Default ports for HTTP and HTTPS
+		return request.protocol === "http" ? 80 : 443;
+	}
+
+	let host = request.headers.host;
+	if (!host || Array.isArray(host)) {
+		return defaultPort();
+	}
+
+	// IPv6 literal support
+	let offset = host[0] === "[" ? host.indexOf("]") + 1 : 0;
+	let index = host.indexOf(":", offset);
+	if (index !== -1) {
+		return parseInt(host.substring(index + 1), 10);
+	}
+	else {
+		return defaultPort();
+	}
+}
+
+export function createLink(request: express.Request, link: string): string {
```

Comment:
The issue is this is being called as part of `authenticateWithRedirect` when going to `/`, if we try to verify that hostname then we'll be stuck in a loop (since the hostname validation needs to access paths after `/` and `/` would try to get validated again.) The biggest security risk that I see is if someone gets in the way of traffic and redirects the user to something else by setting the 'host' header. 

BUT if someone is already mangling traffic and breaking SSL then they can do a redirect in easier ways.